Add a tab showing the production summary

veger opened the pull request:
I tried to reuse the existing UI of YAFC, so I based the whole thing on a `ProjectPageContents` and a `ProjectPageView`, so it would be recognized and fit into the `MainScreen` (tabs) and serialization.
The UI of the new 'Summary' tab is ugly, I tried making it nicer, but I found it hard to understand the `ImGui` and its features and how to use them... So I guess this could be improved separately from this PR?

![YAFC-production-summary](https://user-images.githubusercontent.com/171827/147891150-b04e10a6-6088-4759-8299-f13449a5350f.png)

I also fixed some issues I found while implementing the Summary tab:
* calculating `DatGrid` width when header is not used (I found the header useless here)
* Don't use 'weird' output of `EncapsulateRect` (at least I don't understand the output of this function...) as `lastRect`, but the  actual `rect` of the (old) state. (So the `DataGrid` size can be properly propagated to the `ScrollArea`
* Fixed use `ScrollArea` naming and exposed the parameters to enable the horizontal scrollbar. (:warning: the horizontal scrollbar is not clickable, off by `1.0f` on vertical axis in the click detection, which I was not able to fix...)
* Some other (minor) things

Related to #123

veger committed:
Add Summary tab (with an empty page)

veger committed:
Also use 'over color' when defalt background is used

veger committed:
Show link and flow information in SummaryView

veger committed:
Align goods underneath eachother and mark linked goods when there is not enough production

veger committed:
Hide columns that have no problems (produced/provided == consumed/requested)

veger committed:
Improve to match UI better ('YAFC rounding' and float rounding)

veger committed:
Also calculate width when header is not rendered

veger committed:
(minor) fixes

* Don't use 'weird' output of EncapsulateRect as 'lastRect', but the
  actual rect of the (old) state
* Improve ScrollArea class names to reflect horizontal scrolling as
  well, including a constructor parameter to enable horizontal scrolling
* Renamed some variables
* Typos

veger committed:
Make Summary table scrollable (horizontal and vertical) and add proper header/title text

Note that the horizontal scrollbar is not clickable (ctrl+scroll and
keyboard scrolling is working)

veger committed:
Use event system to only recalculate/prepare when needed

veger committed:
Add checkbox to only show goods with 'issues' (and fix width calculation of table)

ShadowTheAge commented:
Thanks for the contribution

Yafc (with its serialization, etc) was made with possible plugins in mind and this is a good candidate for a plugin (as is: It doesn't change much, mostly adding new stuff)

It is hard to review what has to be changed because all of the formatting noise (clearly my formatter is configured differently)

ShadowTheAge commented:
Also, I think instead of automatically reading all the tabs, tabs should be manually added to this summary page. This way you can have multiple summaries, and also there will be no "test" tabs there, and also you can add a multiplier to a page, as in "how much of these is built".

ShadowTheAge commented:
Also all the calculations should be done on page activation, not on project loading. So, if this page needs all other pages to be solved, it should run these solvers in its own "solve" instead of solving all the pages on a project load.

veger commented:
> It is hard to review what has to be changed because all of the formatting noise (clearly my formatter is configured differently)

I guess so... :cry:
GitHub diffs can ignore whitespace changes with `?w=1` -> https://github.com/ShadowTheAge/yafc/pull/145/files?w=1
That reduces most of the noise.

> Also, I think instead of automatically reading all the tabs, tabs should be manually added to this summary page. This way you can have multiple summaries, and also there will be no "test" tabs there, and also you can add a multiplier to a page, as in "how much of these is built".

Agreed, this was/is on my to-do list, but I thought this PR is already big enough as it is. (I guess it mainly requires some UI to select pages to be included)

> Also all the calculations should be done on page activation, not on project loading. So, if this page needs all other pages to be solved, it should run these solvers in its own "solve" instead of solving all the pages on a project load.

Makes sense, this will be more easy to do when `Summary` contains a list of guids of the pages it should show... So again, it will increase the PR by quite a bit

veger commented on YAFCui/ImGui/ImGuiLayout.cs:
```diff
@@ -240,7 +241,10 @@ public void Dispose()
                 if (hasContent)
-                    gui.lastRect = gui.state.EncapsulateRect(rect);
+                {
+                    gui.state.EncapsulateRect(rect);
+                    gui.lastRect = rect;
```
This breaks the `ErrorRow` of the ProjectPageView...
(without this change the size of the `ScrollArea` is set to the size of the (remaining) window (too small) and scrollbars are never shown...

@ShadowTheAge Any ideas how to fix either?

veger commented on YAFCui/ImGui/ImGuiLayout.cs:
```diff
@@ -240,7 +241,10 @@ public void Dispose()
                 rect.Width += (padding.left + padding.right);
                 rect.Height += (padding.top + padding.bottom);
                 if (hasContent)
-                    gui.lastRect = gui.state.EncapsulateRect(rect);
+                {
+                    gui.state.EncapsulateRect(rect);
+                    gui.lastRect = rect;
+                }
```
This breaks the `ErrorRow` of the ProjectPageView...
(without this change the size of the `ScrollArea` is set to the size of the (remaining) window (too small) and scrollbars are never shown...

@ShadowTheAge Any ideas how to fix either?

veger replied:
ok, found a fix that seems to work for both `ErrorRow` and `ScrollArea`

veger committed:
Keep original lastRect to fix some drawing glitches (e.g. in ErrorRow)

Adding a lastContentRect which contains the (correct/out-of-window)
size, fixes the contentSize calculation in BuildGui().

I did not see any glich(es) anymore, and the ScrollArea size is set
correctly now (showing scrollbars if needed)

veger commented:
By spending another 2 hours on the ImGui to figure it out to solve the issue I found above, I decided not to add the 'manually add tabs to a summary' idea. It will take me loads of time due to my struggle understanding the ImGui class. (I already found this will building this feature, and while I tried to beautify the GUI of the Summary tab).

For me I can select the tabs by having them in the tab bar or not and 1 Summary is enough to align inputs and outputs of the different tabs. So for me this PR serves my (personal) purpose. And spending ~a week of my free time on this is enough... (the factory did not grow at all during this time :disappointed: )

So feel free to accept/deny this PR (I am still willing to fix issues, solve bugs, etc to get it accepted), but feature-wise this is my limit :wink:

veger committed:
Calculate required amount when click on a link icon

veger committed:
Support the search box (ctrl+F)

ShadowTheAge commented:
I will keep this pr open, when I'll add basic support for plugins I'll convert it to a plugin

veger committed:
Adjust Summary View to window height

It is a bit hackish but it seems to work (after updating the UI to
trigger some 'recalculation')

Fixes: #2

veger committed:
Merge branch 'master' into summary-tab

veger committed:
Fix error when SummaryView is set as secondary screen
